storage/range.go

```diff
@@ -822,119 +858,64 @@ func (r *Range) applyRaftCommand(ctx context.Context, index uint64, originNode p
-		// signaling cmd.done below, or the waiting RPC handler might proceed
-		// before we've updated its reply.
-		//
-		// It is important that the error is set before the reply is saved into
-		// the response cache.
-		reply.Header().SetGoError(rErr)
-
-		if proto.IsWrite(args) {
-			// No matter the result, add result to the response cache if this
-			// is a write method. This must be done as part of the execution of
-			// raft commands so that every replica maintains the same responses
-			// to continue request idempotence, even if leadership changes.
-			if err := r.respCache.PutResponse(args.Header().CmdID, reply); err != nil {
-				rErr = newReplicaCorruptionError(
-					util.Errorf("could not put to response cache"), err, rErr)
-				return
-			}
-		}
-	}()
-
-	header := args.Header()
+		return batch, r.newNotLeaderError(lease, originNode)
+	}
 
 	// Check the response cache to ensure idempotency.
 	if proto.IsWrite(args) {
-		if ok, err := r.respCache.GetResponse(header.CmdID, reply); ok && err == nil {
+		if ok, err := r.respCache.GetResponse(batch, args.Header().CmdID, reply); ok && err == nil {
 			if log.V(1) {
 				log.Infoc(ctx, "found response cache entry for %+v", args.Header().CmdID)
 			}
 			// We successfully read from the response cache, so return whatever error
 			// was present in the cached entry (if any).
-			return reply.Header().GoError()
+			return batch, reply.Header().GoError()
 		} else if ok && err != nil {
-			return newReplicaCorruptionError(
-				util.Errorf("could not read from response cache"), err)
+			return batch, newReplicaCorruptionError(util.Errorf("could not read from response cache"), err)
 		}
 	}
 
-	// Create a new batch for the command to ensure all or nothing semantics.
-	batch := r.rm.Engine().NewBatch()
-	defer batch.Close()
-
-	// Create a engine.MVCCStats instance.
-	ms := engine.MVCCStats{}
-
-	// Execute the command; the error will also be set in the reply header.
+	// Execute the command.
+	intents, rErr := r.executeCmd(batch, ms, args, reply)
+	// ALWAYS set the reply header error to the error returned by the
+	// helper. This is the definitive result of the execution. The
+	// error must be set before saving to the response cache.
 	// TODO(tschottdorf,tamird) For #1400, want to refactor executeCmd to not
 	// touch the reply header's error field.
-	intents, err := r.executeCmd(batch, &ms, args, reply)
-	// If the execution of the command wasn't successful, stop here.
-	if err != nil {
-		return err
-	}
-
-	if oldIndex := atomic.LoadUint64(&r.appliedIndex); oldIndex >= index {
-		return newReplicaCorruptionError(
-			util.Errorf("applied index moved backwards: %d >= %d", oldIndex, index))
-	}
-
-	// Advance the applied index atomically within the batch.
-	if err := setAppliedIndex(batch, r.Desc().RaftID, index); err != nil {
-		return newReplicaCorruptionError(
-			util.Errorf("could not update applied index"), err)
-	}
-
+	reply.Header().SetGoError(rErr)
```

Comment:
this needs to always happen, even if we return early (l.862, cp #1519). Probably best to move this to `applyRaftCommand`.